### PR TITLE
Lucas tests#42

### PR DESCRIPTION
**Pull request by carlschroedl**
This pull request implements MarginCalculator integration tests originally defined in this Google Sheet:
https://docs.google.com/spreadsheets/d/1634wP6-N8GG2Fig-yjIOk7vPBn4AijXOrjq6Z2T1K8M/edit?usp=sharing

Some are Lucas' scenarios (#42). Others are from Tideman's 1987 paper.

This pull request only tests MarginCalculator. A subsequent pull request will need to address RankedPairsCalculator. Fortunately, that will be a much more concise set of tests that can rely on the existing TestScenario classes.

This pull request checks in generated code, and the script used to generate it. This is technical debt that needs to be paid by implementing #27 #28 and #29. Once those issues are implemented, we will be able to encode our test input and expected output in more straightforward text formats. At that time we can remove the unhelpful generated code and the code generator. In exchange for the long-term risk of messy test data, we get short-term assurance that margin calculation -- the foundation of this library -- is working correctly in non-trivial scenarios. Since our deadlines are fast-approaching, I think this an acceptable tradeoff.

### CODE REVIEW

File: tests/MarginCalculatorIntegrationTest.php [hunk 1]
```diff
@@ -0,0 +1,272 @@
+<?php
+
+namespace PivotLibre\Tideman;
+
+use PivotLibre\Tideman\TestScenario1;
+use PivotLibre\Tideman\MarginCalculator;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
```
Comment (carlschroedl) [inline tests/MarginCalculatorIntegrationTest.php:9]: This class is probably the only one worth reviewing. I'm particularly interested if you think the doc block describes how the tests were created with sufficient clarity.

File: tests/parse_ballot.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+import sys
```
Comment (carlschroedl) [inline tests/parse_ballot.py:1]: This is the script used to generate the TestScenario classes. It's probably not worth reviewing. Generating php via python? Gross, I know. See PR description for justification of this technical debt.

File: tests/MarginCalculatorIntegrationTest.php [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        foreach ($expectedMargins as $expectedMargin) {
+            $winner = $expectedMargin->getWinner();
+            $loser = $expectedMargin->getLoser();
+            $marginAsString = "{$winner->getId()} -> {$loser->getId()}";
+            $actualMargin = $marginRegistry->get($winner, $loser);
+            $this->assertEquals($expectedMargin, $actualMargin, $marginAsString);
+        }
+    }
+
+    public function testScenario1() : void
```
Comment (carlschroedl) [inline tests/MarginCalculatorIntegrationTest.php:45]: I asked Lucas about this. No response yet.

Comment (carlschroedl) [reply tests/MarginCalculatorIntegrationTest.php:45]: Resolved. Lucas independently confirmed my suspicion -  that the original margin values in the Scenario 1 worksheet were double what they should have been. He found the same problem in the Scenario 5 worksheet, which is not part of this pull request. The margin values for Scenario 1 in this pull request are half of the original values in the worksheet, and the tests pass.  w00t!

**Review by andrewshell — APPROVED** [review]

**Pull request merged**